examples/stm32/nucleo-h743zi-baremetal/main.c

```diff
@@ -1,67 +1,74 @@
-void SysTick_Handler(void) { s_ticks++; }
-uint64_t mg_millis(void) { return s_ticks; }
+#define LED LED2              // Use blue LED for blinking
+#define BLINK_PERIOD_MS 1000  // LED blinking period in millis
 
-static void timer_cb(void *arg) {
-  gpio_toggle(LED2);                      // Blink LED
-  MG_INFO(("ticks: %lld", mg_millis()));  // Log something
-  (void) arg;
+static uint64_t s_ticks;      // Milliseconds since boot
+void SysTick_Handler(void) {  // SyStick IRQ handler, triggered every 1ms
+  s_ticks++;
 }
 
-int main(void) {
-  clock_init();
-  systick_init(CPU_FREQUENCY / 1000);  // Increment s_ticks every ms
-  gpio_output(LED2);                   // Setup LED
-  uart_init(UART_DEBUG, 115200);       // Initialise debug printf
+uint64_t mg_millis(void) {  // Let Mongoose use our uptime function
+  return s_ticks;           // Return number of milliseconds since boot
+}
 
-  char rev = chiprev();
-  MG_INFO(("Chip revision: %c, max cpu clock: %u MHz", rev, (rev == 'V') ? 480 : 400));
+void mg_random(void *buf, size_t len) {  // Use on-board RNG
+  for (size_t n = 0; n < len; n += sizeof(uint32_t)) {
+    uint32_t r = rng_read();
+    memcpy((char *) buf + n, &r, n + sizeof(r) > len ? len - n : sizeof(r));
+  }
+}
 
+static void timer_fn(void *arg) {
+  gpio_toggle(LED);                               // Blink LED
+  struct mg_tcpip_if *ifp = arg;                  // And show
+  const char *names[] = {"down", "up", "ready"};  // network stats
+  MG_INFO(("Ethernet: %s, IP: %M, rx:%u, tx:%u, dr:%u, er:%u",
+           names[ifp->state], mg_print_ip4, &ifp->ip, ifp->nrecv, ifp->nsent,
+           ifp->ndrop, ifp->nerr));
+}
 
-  // Initialise Ethernet. Enable MAC GPIO pins, see UM2407 for MB1364 boards
-  // section 6.6.7:
-  // https://www.st.com/resource/en/user_manual/um2407-stm32h7-nucleo144-boards-mb1364-stmicroelectronics.pdf
+static void ethernet_init(void) {
+  // Initialise Ethernet. Enable MAC GPIO pins, see
+  // https://www.farnell.com/datasheets/2014265.pdf section 6.10
   uint16_t pins[] = {PIN('A', 1),  PIN('A', 2),  PIN('A', 7),
                      PIN('B', 13), PIN('C', 1),  PIN('C', 4),
                      PIN('C', 5),  PIN('G', 11), PIN('G', 13)};
   for (size_t i = 0; i < sizeof(pins) / sizeof(pins[0]); i++) {
     gpio_init(pins[i], GPIO_MODE_AF, GPIO_OTYPE_PUSH_PULL, GPIO_SPEED_INSANE,
-              GPIO_PULL_NONE, 11);
+              GPIO_PULL_NONE, 11);  // 11 is the Ethernet function
   }
-  nvic_enable_irq(61);                          // Setup Ethernet IRQ handler
-  RCC->APB4ENR |= BIT(1);                       // Enable SYSCFG
+  NVIC_EnableIRQ(ETH_IRQn);                     // Setup Ethernet IRQ handler
   SETBITS(SYSCFG->PMCR, 7 << 21, 4 << 21);      // Use RMII (12.3.1)
   RCC->AHB1ENR |= BIT(15) | BIT(16) | BIT(17);  // Enable Ethernet clocks
-  RCC->AHB1RSTR |= BIT(15);                     // ETHMAC force reset
-  RCC->AHB1RSTR &= ~BIT(15);                    // ETHMAC release reset
```

Comment:
We used to reset the MAC controller

Reply:
Do we need that ? Isn't MAC controller reset after a reboot?

Reply:
Yes, it should, but we've been doing it, I thought it was for consistency. No resets from now on